devito/symbolics/search.py

```diff
@@ -40,98 +46,67 @@ def __init__(self, query, mode, deep=False):
         ----------
         query
             Any query from :mod:`queries`.
-        mode : str
-            Either 'unique' or 'all' (catch all instances).
+        order : str
+            Either `preorder` or `postorder`, for the search order.
         deep : bool, optional
             If True, propagate the search within an Indexed's indices. Defaults to False.
         """
         self.query = query
-        self.collection = self.modes[mode]
+        self.order = order
         self.deep = deep
 
-    def _next(self, expr):
+    def _next(self, expr) -> Iterator[sympy.Basic]:
```

Comment:
Some of the args and/or indices might not be sympy.Basic, it could be for example some plain python or numpy numbers.